src/ViewHelpers/AliasViewHelper.php

```diff
@@ -78,14 +78,23 @@ public function initializeArguments()
     public static function renderStatic(array $arguments, \Closure $renderChildrenClosure, RenderingContextInterface $renderingContext)
     {
         $templateVariableContainer = $renderingContext->getVariableProvider();
+        $originalVariableContainer = clone $templateVariableContainer;
```

Comment:
How about `$renderingContext->setVariableProvider($originalVariableContainer)` afterwards instead?

Reply:
This is what the additional tests I just added cover: If you use `f:variable` inside `f:alias`, those variables imho shouldn't be reset (and this would be breaking).

Reply:
This is getting ridiculously complicated though:

```html
<f:variable name="foo" value="1"/>
<f:alias map="{foo: 2}">
    <f:variable name="foo" value="3"/>
</f:alias>
```
What should be the value of `foo` after this?

Reply:
In my opinion, the rule should be: ViewHelpers injecting variables into their children (like `f:alias` and `f:for`) create local variables, all others are global.

I recognize that this gets complicated, but `f:for` and `f:variable` are out there... Do you think that the current alternative that creates side effects is better/easier?

Reply:
I'm thinking about how to actually solve this. The example I just gave is hard to implement as expected. The `f:alias` viewhelper can only restore `foo` to the state it had before it was rendered. This way it overwrites the value set by the 2nd `f:variable` viewhelper.

So a possible workflow:

1. Back up variables before rendering
2. Render
3. Go through all `map` keys
4. If a `map` value is different from the one in the `map`, keep it
5. If a `map` value is the same as the one in the `map` and a matching variable is present in the backup, restore the backup value
6. Otherwise remove the variable again

Reply:
If I understood this correctly, this would lead to the following rule:

`f:variable` (and other potential ViewHelpers that create/modify variables directly, something like `f:array.push`) always create global variables.

Reply:
I'd avoid the term "global variables" until Fluid actually has variable scopes. But yeah, unless variables are created by a viewhelper for child rendering, they will stay around after rendering.

Reply:
Maybe the implementation would be clearer if there would be a special VariableProvider implementation for these circumstances?

It could encapsulate two StandardVariableProviders ("global/outside" and "local/inside"). `get()` would check local variables first and fall back to global variables. `add()` would always write to global variables. Before `renderChildren()`, the new VariableProvider would be constructed, afterwards it could just be reset to the global VariableProvider.

Reply:
That's basically implementing variable scopes in Fluid. ;-)

Small correction: `add()` would always write to local variables here, that's what `f:alias` is expected to do. Children rendered within would access the "global" provider instead.

Reply:
I've created a separate VariableProvider that can be used by all ViewHelpers that create local variables. This PR now depends on a pre-patch #829 that removes the constructor from the underlying interface.